### DISCUSSION TITLE
There is no upper limit on the number of messages buffer for a sleep end device.

### DISCUSSION THREAD

**Discussion by jinpeng1989** [openthread/openthread#8084] [Q&A]
The timeout of our sleep-end-device is 1800 seconds. When a SED leaves the network. All the messages send to it will be buffer until 
timeout expired. The disconnected device may have consumed all of the messages buffer, other devices cannot work properly.

**To Reproduce** Information to reproduce the behavior, including:
1. Git commit id: 3a98779dc9f0875af9c7b137e764c9663bfdc9c3
    ot-br-posix/2022-06-02-3a98779-2; POSIX; Feb 15 2021 15:22:37
3. IEEE 802.15.4 hardware platform: SL-OPENTHREAD/2.1.0.0_GitHub-8f92d2dc8_v1; EFR32; Jul 19 2022 14:18:51
4. Build steps
    Join a SED(timeout=1800 second). Power off SED, than send message to this SED. The otbr-agent buffer will continue to decrease until it runs out.

**Expected behavior** A clear and concise description of what you expected to happen.
There is a maximum of messages buffer for a given sleep end device.  Messages exceeding the upper limit can be dropped.

![image](https://user-images.githubusercontent.com/97089879/186648389-7096fad0-2fde-41d5-9d19-ab3bc1bf367c.png)
![image](https://user-images.githubusercontent.com/97089879/186648499-c5e9e858-2b35-45d5-a8ce-ffc3c24cc8fc.png)

**Comment by jwhui** [comment]
What you are observing is expected. If OpenThread has buffers available, it will utilize them. However, if those buffers are needed at a later time, it can reclaim them by evicting already-buffered messages.

See `MessagePool::NewBuffer()` call to `ReclaimBuffers()`:
https://github.com/openthread/openthread/blob/cba1bebf1f1ef3531bfe5397c75c3c6e045218b6/src/core/common/message.cpp#L122

Which calls `MeshForwarder::EvictMessage()`:
https://github.com/openthread/openthread/blob/cba1bebf1f1ef3531bfe5397c75c3c6e045218b6/src/core/thread/mesh_forwarder_ftd.cpp#L205

As a result, just because the router is buffering messages for a child, that should not prevent it from buffering other messages that it needs to forward later on.

**Reply by jinpeng1989** [reply]
So only when there is no buffer can be reclaim means that no available message buffer. If the level of this log can be improved?

If define OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT to 8. Whether otbr-agent can guarantee 8 buffer for a given SED? Will the newest or the oldest be dropped?

**Reply by jwhui** [reply]
> So only when there is no buffer can be reclaim means that no available message buffer. If the level of this log can be improved?

If no buffers can be reclaimed allocation will fail. Can you point to a specific log line that you are referring to?

> If define OPENTHREAD_CONFIG_DEFAULT_SED_DATAGRAM_COUNT to 8. Whether otbr-agent can guarantee 8 buffer for a given SED? Will the newest or the oldest be dropped?

The current implementation will evict the message with lowest priority. If multiple messages exist at the lowest priority, it will evict the oldest one. See implementation at:
https://github.com/openthread/openthread/blob/cba1bebf1f1ef3531bfe5397c75c3c6e045218b6/src/core/thread/mesh_forwarder_ftd.cpp#L205

**Reply by jinpeng1989** [reply]
The free number is greater than total in our test environment. Please help to check.
![bufferinfo_free_number_error](https://user-images.githubusercontent.com/97089879/186827268-d425ff06-4a99-4c3b-a197-899a9c0306a8.jpg)

**Reply by jwhui** [reply]
Number of free buffers should not be greater than number of total buffers.

Number of total buffers should be static - see implementation:
https://github.com/openthread/openthread/blob/ba006225e03c57aeb09599b5acb67e09eb98ee95/src/core/common/message.cpp#L181

Your [original post](https://github.com/openthread/openthread/discussions/8084#discussion-4333799) showed 16383 total buffers. However, the post immediately above shows 8192 buffers. Are these difference devices? If so, can you provide more information on how their configurations differ?

**Reply by jinpeng1989** [reply]
These are two different devices. The total buffers of the first device is 16383. The second device is 8192. Only this configurations differ.

**Comment by jinpeng1989** [comment]
> If no buffers can be reclaimed allocation will fail. Can you point to a specific log line that you are referring to?

    [https://github.com/openthread/openthread/blob/main/src/core/common/message.cpp#L134](url)

**Reply by jwhui** [reply]
This log line: https://github.com/openthread/openthread/blob/ba006225e03c57aeb09599b5acb67e09eb98ee95/src/core/common/message.cpp#L134

Indicates that `ReclaimBuffers()` could not provide any free buffers.
